<a href="https://colab.research.google.com/github/Pratyush2722/100daysML/blob/main/Hackathon.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Importing Libraries -- Start**

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

**Load Dataset**

In [2]:
train = pd.read_csv("/content/sample_data/train.csv")
test = pd.read_csv("/content/sample_data/test.csv")

Lets Analyse the data

In [3]:
train.head()

,id,gender,area,qualification,income,marital_status,vintage,claim_amount,num_policies,policy,type_of_policy,cltv
0,1,Male,Urban,Bachelor,5L-10L,1,5,5790,More than 1,A,Platinum,64308
1,2,Male,Rural,High School,5L-10L,0,8,5080,More than 1,A,Platinum,515400
2,3,Male,Urban,Bachelor,5L-10L,1,8,2599,More than 1,A,Platinum,64212
3,4,Female,Rural,High School,5L-10L,0,7,0,More than 1,A,Platinum,97920
4,5,Male,Urban,High School,More than 10L,1,6,3508,More than 1,A,Gold,59736


This shows the first 5 rows of the training data

**Check the shape of the data**

In [4]:
train.shape

(89392, 12)

It shows us that data has  89392 rows and 12 columns


**Understanding columns**

In [5]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89392 entries, 0 to 89391
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id              89392 non-null  int64 
 1   gender          89392 non-null  object
 2   area            89392 non-null  object
 3   qualification   89392 non-null  object
 4   income          89392 non-null  object
 5   marital_status  89392 non-null  int64 
 6   vintage         89392 non-null  int64 
 7   claim_amount    89392 non-null  int64 
 8   num_policies    89392 non-null  object
 9   policy          89392 non-null  object
 10  type_of_policy  89392 non-null  object
 11  cltv            89392 non-null  int64 
dtypes: int64(5), object(7)
memory usage: 8.2+ MB


Thhis data shows us the data types , missing values and info about the columns

***Statistical Understanding***

In [6]:
train.describe()

,id,marital_status,vintage,claim_amount,cltv
count,89392.000000,89392.000000,89392.000000,89392.000000,89392.000000
mean,44696.500000,0.575488,4.595669,4351.502416,97952.828978
std,25805.391969,0.494272,2.290446,3262.359775,90613.814793
min,1.000000,0.000000,0.000000,0.000000,24828.000000
25%,22348.750000,0.000000,3.000000,2406.000000,52836.000000
50%,44696.500000,1.000000,5.000000,4089.000000,66396.000000
75%,67044.250000,1.000000,6.000000,6094.000000,103440.000000
max,89392.000000,1.000000,8.000000,31894.000000,724068.000000


This gives the min , max and count of the  columns . This helps us to find the weird values ,outliers , huge ranges in the data .

**Missing values check**

In [7]:
train.isnull().sum()

,0
id,0
gender,0
area,0
qualification,0
income,0
marital_status,0
vintage,0
claim_amount,0
num_policies,0
policy,0


This shows us that there are no missing values in any columns otherwise we would have used methods to deal with them.

***Separate Features and Target***

In [8]:
X = train.drop("cltv", axis=1)
y = train["cltv"]

we are assigning input features in X-axis and CLTV as output in Y-axis

***Train Test Split***

In [9]:
X_train, X_valid, y_train, y_valid = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

We are splitting the data into parts as we have to test the our model for ourselves , then we can test it on the test data


***Encoding Categorical Columns***

In [10]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

X_train["gender"] = le.fit_transform(X_train["gender"])

*We have to encode this columns as model only understand numbers.*

***Linear Regression ***

First Lets take a Simple Model

In [11]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()

***Training the model***

In [12]:
X_train.select_dtypes(include="object").columns

Index(['area', 'qualification', 'income', 'num_policies', 'policy',
       'type_of_policy'],
      dtype='object')

In [13]:
X_train[['income', 'num_policies']].head()

,income,num_policies
22085,More than 10L,More than 1
19413,5L-10L,1
30309,<=2L,1
333,5L-10L,1
60407,5L-10L,More than 1


In [14]:
X_train[['income', 'num_policies']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 71513 entries, 22085 to 15795
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   income        71513 non-null  object
 1   num_policies  71513 non-null  object
dtypes: object(2)
memory usage: 1.6+ MB


In [15]:
cat_cols = X_train.select_dtypes(include="object").columns

print(cat_cols)

Index(['area', 'qualification', 'income', 'num_policies', 'policy',
       'type_of_policy'],
      dtype='object')


In [16]:
from sklearn.preprocessing import LabelEncoder

for col in cat_cols:
    le = LabelEncoder()

    X_train[col] = le.fit_transform(X_train[col])
    X_valid[col] = le.transform(X_valid[col])

In [17]:
model.fit(X_train, y_train)

LinearRegression()

In [18]:
X_valid.select_dtypes(include="object").columns

Index(['gender'], dtype='object')

In [19]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

# Retrieve original gender column for X_train and X_valid using their indices
# This ensures the LabelEncoder is fitted on string data
original_X_train_gender = train.loc[X_train.index, "gender"]
original_X_valid_gender = train.loc[X_valid.index, "gender"]

# Fit the encoder on the original string values from the training data
le.fit(original_X_train_gender)

# Transform both train and valid gender columns using the fitted encoder
X_train["gender"] = le.transform(original_X_train_gender)
X_valid["gender"] = le.transform(original_X_valid_gender)

In [20]:
X_valid.select_dtypes(include="object").columns

Index([], dtype='object')

In [21]:
preds = model.predict(X_valid)

In [22]:
from sklearn.metrics import r2_score

score = r2_score(y_valid, preds)

print(score)

0.15078321465112854


Simple model - Linear Regression to test the data . It doesnt fit well as data is  having non-linear relationships .

***CATBOOST MODELTESTING***

In [23]:

from catboost import CatBoostRegressor

In [24]:
model = CatBoostRegressor(
    iterations=500,
    learning_rate=0.05,
    depth=6,
    verbose=0
)

In [25]:
model.fit(X_train, y_train)

CatBoostRegressor(depth=6, iterations=500, learning_rate=0.05, loss_function='RMSE', verbose=0)

In [26]:
preds = model.predict(X_valid)

In [27]:
score2 = r2_score(y_valid, preds)

print(score2)

0.15545752151170789


Loading dataset again

In [28]:
train = pd.read_csv("/content/sample_data/train.csv")
test = pd.read_csv("/content/sample_data/test.csv")

In [29]:
X = train.drop(["cltv", "id"],axis=1)
y = train["cltv"]

X_train, X_valid, y_train, y_valid = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [30]:
income_map = {
    "<=2L": 1,
    "2L-5L": 2,
    "5L-10L": 3,
    "More than 10L": 4
}

In [31]:
X_train["income_encoded"] = X_train["income"].map(income_map)
X_valid["income_encoded"] = X_valid["income"].map(income_map)

In [32]:
policy_map = {
    "1": 1,
    "More than 1": 2
}

In [33]:
X_train["policy_count_encoded"] = X_train["num_policies"].map(policy_map)
X_valid["policy_count_encoded"] = X_valid["num_policies"].map(policy_map)

In [34]:
# Ensure income_encoded and policy_count_encoded are present in case X_train/X_valid were reset
X_train["income_encoded"] = X_train["income"].map(income_map)
X_valid["income_encoded"] = X_valid["income"].map(income_map)

X_train["policy_count_encoded"] = X_train["num_policies"].map(policy_map)
X_valid["policy_count_encoded"] = X_valid["num_policies"].map(policy_map)

X_train["claim_per_policy"] = (
    X_train["claim_amount"] /
    X_train["policy_count_encoded"]
)

X_valid["claim_per_policy"] = (
    X_valid["claim_amount"] /
    X_valid["policy_count_encoded"]
)

In [35]:
X_train["income_vintage"] = (
    X_train["income_encoded"] *
    X_train["vintage"]
)

X_valid["income_vintage"] = (
    X_valid["income_encoded"] *
    X_valid["vintage"]
)

In [36]:
X_train.head()

,gender,area,qualification,income,marital_status,vintage,claim_amount,num_policies,policy,type_of_policy,income_encoded,policy_count_encoded,claim_per_policy,income_vintage
22085,Male,Urban,Others,More than 10L,1,6,2071,More than 1,A,Gold,4,2,1035.5,24
19413,Male,Urban,Others,5L-10L,1,3,7275,1,B,Platinum,3,1,7275.0,9
30309,Female,Urban,High School,<=2L,1,6,4716,1,A,Platinum,1,1,4716.0,6
333,Female,Urban,High School,5L-10L,0,6,0,1,A,Platinum,3,1,0.0,18
60407,Female,Urban,Bachelor,5L-10L,1,5,0,More than 1,B,Silver,3,2,0.0,15


In [37]:
print(X_train.columns)

Index(['gender', 'area', 'qualification', 'income', 'marital_status',
       'vintage', 'claim_amount', 'num_policies', 'policy', 'type_of_policy',
       'income_encoded', 'policy_count_encoded', 'claim_per_policy',
       'income_vintage'],
      dtype='object')


In [38]:
cat_cols = X_train.select_dtypes(include="object").columns.tolist()

print(cat_cols)

['gender', 'area', 'qualification', 'income', 'num_policies', 'policy', 'type_of_policy']


In [39]:
from catboost import CatBoostRegressor

model = CatBoostRegressor(
    iterations=1000,
    learning_rate=0.03,
    depth=8,
    loss_function='RMSE',
    eval_metric='R2',
    verbose=100
)

In [40]:
model.fit(
    X_train,
    y_train,
    cat_features=cat_cols,
    eval_set=(X_valid, y_valid),
    early_stopping_rounds=100
)

0:	learn: 0.0088229	test: 0.0088497	best: 0.0088497 (0)	total: 132ms	remaining: 2m 11s
100:	learn: 0.1637985	test: 0.1599306	best: 0.1599306 (100)	total: 7.4s	remaining: 1m 5s
200:	learn: 0.1683978	test: 0.1603107	best: 0.1603508 (180)	total: 15.2s	remaining: 1m
Stopped by overfitting detector  (100 iterations wait)

bestTest = 0.1603508484
bestIteration = 180

Shrink model to first 181 iterations.


CatBoostRegressor(depth=8, eval_metric='R2', iterations=1000, learning_rate=0.03, loss_function='RMSE', verbose=100)

In [41]:
preds = model.predict(X_valid)

In [42]:
score = r2_score(y_valid, preds)

print(score)

0.16035084840202396


K-Fold Cross Validation

In [43]:
from sklearn.model_selection import KFold

Create Folds

In [44]:
kf = KFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

In [45]:
print(X_train.columns)

Index(['gender', 'area', 'qualification', 'income', 'marital_status',
       'vintage', 'claim_amount', 'num_policies', 'policy', 'type_of_policy',
       'income_encoded', 'policy_count_encoded', 'claim_per_policy',
       'income_vintage'],
      dtype='object')


In [46]:
scores = []

for train_idx, valid_idx in kf.split(X):

    X_train = X.iloc[train_idx]
    X_valid = X.iloc[valid_idx]

    y_train = y.iloc[train_idx]
    y_valid = y.iloc[valid_idx]

    model = CatBoostRegressor(
        iterations=500,
        learning_rate=0.03,
        depth=8,
        verbose=0
    )

    model.fit(
        X_train,
        y_train,
        cat_features=cat_cols
    )

    preds = model.predict(X_valid)

    score = r2_score(y_valid, preds)

    scores.append(score)

    print(score)

0.1598139436762369
0.15957539355985406
0.15851712873053314
0.16532006556585532
0.15921615354620988


In [47]:
print("Average Score:", sum(scores)/len(scores))

Average Score: 0.16048853701573787


**K Fold Cross Validation**

We will add all the features to X

In [48]:
# ==============================
# IMPORT LIBRARIES
# ==============================

import pandas as pd
import numpy as np

from sklearn.model_selection import KFold
from sklearn.metrics import r2_score

from catboost import CatBoostRegressor


In [49]:

# ==============================
# LOAD DATA
# ==============================

train = pd.read_csv("/content/sample_data/train.csv")
test = pd.read_csv("/content/sample_data/test.csv")


In [50]:
# ==============================
# CREATE X AND y
# ==============================

X = train.drop(["cltv", "id"], axis=1)
y = train["cltv"]

In [51]:

# ==============================
# FEATURE ENGINEERING
# ==============================

# ---- Income Mapping ----

income_map = {
    "<=2L": 1,
    "2L-5L": 2,
    "5L-10L": 3,
    "More than 10L": 4
}

X["income_encoded"] = X["income"].map(income_map)


# ---- Policy Count Mapping ----

policy_map = {
    "1": 1,
    "More than 1": 2
}

X["policy_count_encoded"] = X["num_policies"].map(policy_map)


# ---- Claim Per Policy ----

X["claim_per_policy"] = (
    X["claim_amount"] /
    X["policy_count_encoded"]
)

X["claim_vintage_ratio"] = (
    X["claim_amount"] /
    (X["vintage"] + 1)
)


# ---- Income × Vintage ----

X["income_vintage"] = (
    X["income_encoded"] *
    X["vintage"]
)


In [52]:
# ==============================
# CATEGORICAL COLUMNS
# ==============================

cat_cols = X.select_dtypes(include="object").columns.tolist()

print("Categorical Columns:")
print(cat_cols)

print("\nFinal Features:")
print(X.columns)




Categorical Columns:
['gender', 'area', 'qualification', 'income', 'num_policies', 'policy', 'type_of_policy']

Final Features:
Index(['gender', 'area', 'qualification', 'income', 'marital_status',
       'vintage', 'claim_amount', 'num_policies', 'policy', 'type_of_policy',
       'income_encoded', 'policy_count_encoded', 'claim_per_policy',
       'claim_vintage_ratio', 'income_vintage'],
      dtype='object')


In [53]:
# ==============================
# K-FOLD CROSS VALIDATION
# ==============================

kf = KFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

scores = []

In [54]:
# ==============================
# TRAINING LOOP
# ==============================

for fold, (train_idx, valid_idx) in enumerate(kf.split(X)):

    print(f"\n========== Fold {fold+1} ==========")

    # Split data

    X_train = X.iloc[train_idx]
    X_valid = X.iloc[valid_idx]

    y_train = y.iloc[train_idx]
    y_valid = y.iloc[valid_idx]


    # Create Model

    # model = CatBoostRegressor(
    #     iterations=1000,
    #     learning_rate=0.03,
    #     depth=8,
    #     loss_function='RMSE',
    #     eval_metric='R2',
    #     verbose=100
    # )
    model1 = CatBoostRegressor(
        iterations=1500,
        learning_rate=0.02,
        depth=6,
        loss_function='RMSE',
        eval_metric='R2',
        verbose=100
    )

    model2 = CatBoostRegressor(

    iterations=2000,
    learning_rate=0.015,
    depth=7,

    loss_function='RMSE',
    eval_metric='R2',

    verbose=100
)


    # Train Model

    model1.fit(
        X_train,
        y_train,
        cat_features=cat_cols,
        eval_set=(X_valid, y_valid),
        early_stopping_rounds=100
    )

    model2.fit(
        X_train,
        y_train,
        cat_features=cat_cols,
        eval_set=(X_valid, y_valid),
        early_stopping_rounds=100
    )


    # Predict

    preds1 = model1.predict(X_valid)
    preds2 = model2.predict(X_valid)

    final_preds = (preds1 + preds2) / 2
    # Calculate Score

    score = r2_score(y_valid, final_preds)

    scores.append(score)

    print(f"Fold {fold+1} R2 Score: {score}")



========== Fold 1 ==========
0:	learn: 0.0060326	test: 0.0060163	best: 0.0060163 (0)	total: 153ms	remaining: 3m 48s
100:	learn: 0.1564847	test: 0.1563370	best: 0.1563370 (100)	total: 6.99s	remaining: 1m 36s
200:	learn: 0.1627720	test: 0.1599491	best: 0.1599500 (199)	total: 12s	remaining: 1m 17s
300:	learn: 0.1646765	test: 0.1601364	best: 0.1601438 (287)	total: 18s	remaining: 1m 11s
400:	learn: 0.1658469	test: 0.1601976	best: 0.1602012 (386)	total: 22.3s	remaining: 1m 1s
500:	learn: 0.1669644	test: 0.1602850	best: 0.1602866 (479)	total: 28s	remaining: 55.8s
Stopped by overfitting detector  (100 iterations wait)

bestTest = 0.1602865653
bestIteration = 479

Shrink model to first 480 iterations.
0:	learn: 0.0045503	test: 0.0044830	best: 0.0044830 (0)	total: 121ms	remaining: 4m 1s
100:	learn: 0.1511662	test: 0.1512254	best: 0.1512254 (100)	total: 8.11s	remaining: 2m 32s
200:	learn: 0.1622050	test: 0.1596988	best: 0.1596988 (200)	total: 14.9s	remaining: 2m 13s
300:	learn: 0.1650212	test: 0

In [55]:
# ==============================
# FINAL AVERAGE SCORE
# ==============================

avg_score = np.mean(scores)

print("\n==============================")
print("Average R2 Score:", avg_score)
print("==============================")



Average R2 Score: 0.16074925904771098


In [56]:
# ==============================
# TRAIN FINAL MODEL 1
# ==============================

final_model1 = CatBoostRegressor(
    iterations=1500,
    learning_rate=0.02,
    depth=6,
    loss_function='RMSE',
    eval_metric='R2',
    verbose=100
)

final_model1.fit(
    X,
    y,
    cat_features=cat_cols
)

# ==============================
# TRAIN FINAL MODEL 2
# ==============================

final_model2 = CatBoostRegressor(
    iterations=2000,
    learning_rate=0.015,
    depth=7,
    loss_function='RMSE',
    eval_metric='R2',
    verbose=100
)

final_model2.fit(
    X,
    y,
    cat_features=cat_cols
)

0:	learn: 0.0060193	total: 92.6ms	remaining: 2m 18s
100:	learn: 0.1564251	total: 6.98s	remaining: 1m 36s
200:	learn: 0.1623716	total: 14.5s	remaining: 1m 33s
300:	learn: 0.1640356	total: 19.8s	remaining: 1m 19s
400:	learn: 0.1650284	total: 26.5s	remaining: 1m 12s
500:	learn: 0.1658674	total: 31.3s	remaining: 1m 2s
600:	learn: 0.1671130	total: 38.2s	remaining: 57.2s
700:	learn: 0.1684987	total: 44.2s	remaining: 50.4s
800:	learn: 0.1696565	total: 51.1s	remaining: 44.6s
900:	learn: 0.1707648	total: 57.7s	remaining: 38.3s
1000:	learn: 0.1717929	total: 1m 5s	remaining: 32.7s
1100:	learn: 0.1729676	total: 1m 12s	remaining: 26.1s
1200:	learn: 0.1740025	total: 1m 20s	remaining: 19.9s
1300:	learn: 0.1750684	total: 1m 26s	remaining: 13.2s
1400:	learn: 0.1760918	total: 1m 34s	remaining: 6.67s
1499:	learn: 0.1770802	total: 1m 41s	remaining: 0us
0:	learn: 0.0045884	total: 114ms	remaining: 3m 47s
100:	learn: 0.1506293	total: 9.61s	remaining: 3m
200:	learn: 0.1615720	total: 18.1s	remaining: 2m 41s
30

CatBoostRegressor(depth=7, eval_metric='R2', iterations=2000, learning_rate=0.015, loss_function='RMSE', verbose=100)

In [57]:
# ==============================
# TEST DATA FEATURE ENGINEERING
# ==============================

test_X = test.drop(["id"], axis=1)


# Income Encoding

test_X["income_encoded"] = test_X["income"].map(income_map)


# Policy Count Encoding

test_X["policy_count_encoded"] = test_X["num_policies"].map(policy_map)


# Claim Per Policy

test_X["claim_per_policy"] = (
    test_X["claim_amount"] /
    test_X["policy_count_encoded"]
)


# Income × Vintage

test_X["income_vintage"] = (
    test_X["income_encoded"] *
    test_X["vintage"]
)


In [58]:
# ==============================
# FINAL PREDICTIONS
# ==============================

# Add missing feature to test_X
test_X["claim_vintage_ratio"] = (
    test_X["claim_amount"] /
    (test_X["vintage"] + 1)
)

# ==============================
# TEST PREDICTIONS
# ==============================

preds1 = final_model1.predict(test_X)

preds2 = final_model2.predict(test_X)

# Ensemble Average

final_preds = (preds1 + preds2) / 2

In [59]:
# ==============================
# CREATE SUBMISSION FILE
# ==============================
submission = pd.DataFrame()

submission["id"] = test["id"]

submission["cltv"] = final_preds

In [60]:
# ==============================
# SAVE SUBMISSION
# ==============================

submission.to_csv("submission.csv", index=False)

print("\nSubmission file created successfully!")


Submission file created successfully!


In [61]:
print(type(test))
print(test.shape)

<class 'pandas.core.frame.DataFrame'>
(59595, 11)


In [62]:
print(type(final_preds))
print(len(final_preds))

<class 'numpy.ndarray'>
59595


In [63]:
submission = pd.DataFrame()

submission["id"] = test["id"]

submission["cltv"] = final_preds

print(submission.head())

      id           cltv
0  89393   92410.434109
1  89394  129287.804317
2  89395   91709.080042
3  89396   89139.320529
4  89397  125957.724668


In [64]:
submission.to_csv("submission.csv", index=False)

print("DONE")

DONE


In [65]:
import os

print(os.listdir())

['.config', '.ipynb_checkpoints', 'submission.csv', 'catboost_info', 'sample_data']
